controllers/drcluster_controller.go

```diff
@@ -0,0 +1,297 @@
+		return ctrl.Result{}, fmt.Errorf("failed to handle cluster fencing: %w",
+			u.validatedSetFalse("FencingHandlingFailed", err))
+	}
+
+	// TODO: Setup views for storage class and VRClass to read and report IDs
+
+	return ctrl.Result{}, u.validatedSetTrue("Succeeded", "drcluster validated")
+}
+
+func validateDRCluster(ctx context.Context, drcluster *ramen.DRCluster, apiReader client.Reader,
+	objectStoreGetter ObjectStoreGetter, listKeyPrefix string,
+	log logr.Logger,
+) (string, error) {
+	reason, err := validateS3Profile(ctx, apiReader, objectStoreGetter, drcluster, listKeyPrefix, log)
+	if err != nil {
+		return reason, err
+	}
+
+	err = validateCIDRsFormat(drcluster, log)
+	if err != nil {
+		return ReasonValidationFailed, err
+	}
+
+	// TODO: Validate managedCluster name? and also ensure it is not deleted!
+
+	return "", nil
+}
+
+func validateS3Profile(ctx context.Context, apiReader client.Reader,
+	objectStoreGetter ObjectStoreGetter,
+	drcluster *ramen.DRCluster, listKeyPrefix string, log logr.Logger) (string, error) {
+	if drcluster.Spec.ClusterFence == ramen.ClusterFenceStateFenced ||
+		drcluster.Spec.ClusterFence == ramen.ClusterFenceStateManuallyFenced {
+		return "", nil
+	}
+
+	if reason, err := s3ProfileValidate(ctx, apiReader, objectStoreGetter,
+		drcluster.Spec.S3ProfileName, listKeyPrefix, log); err != nil {
+		return reason, err
+	}
+
+	return "", nil
+}
+
+func s3ProfileValidate(ctx context.Context, apiReader client.Reader,
+	objectStoreGetter ObjectStoreGetter, s3ProfileName, listKeyPrefix string,
+	log logr.Logger,
+) (string, error) {
+	objectStore, err := objectStoreGetter.ObjectStore(ctx, apiReader, s3ProfileName, "drpolicy validation", log)
+	if err != nil {
+		return "s3ConnectionFailed", fmt.Errorf("%s: %w", s3ProfileName, err)
+	}
+
+	if _, err := objectStore.ListKeys(listKeyPrefix); err != nil {
+		return "s3ListFailed", fmt.Errorf("%s: %w", s3ProfileName, err)
+	}
+
+	return "", nil
+}
+
+func validateCIDRsFormat(drcluster *ramen.DRCluster, log logr.Logger) error {
+	// validate the CIDRs format
+	invalidCidrs := []string{}
+
+	for i := range drcluster.Spec.CIDRs {
+		if _, _, err := net.ParseCIDR(drcluster.Spec.CIDRs[i]); err != nil {
+			invalidCidrs = append(invalidCidrs, drcluster.Spec.CIDRs[i])
+
+			log.Error(err, ReasonValidationFailed)
+		}
+	}
+
+	if len(invalidCidrs) > 0 {
+		return fmt.Errorf("invalid CIDRs specified %s", strings.Join(invalidCidrs, ", "))
+	}
+
+	return nil
+}
+
+type drclusterUpdater struct {
+	ctx    context.Context
+	object *ramen.DRCluster
+	client client.Client
+	log    logr.Logger
+}
+
+func (u *drclusterUpdater) validatedSetTrue(reason, message string) error {
+	return u.statusConditionSet(ramen.DRClusterValidated, metav1.ConditionTrue, reason, message)
+}
+
+func (u *drclusterUpdater) validatedSetFalse(reason string, err error) error {
+	if err1 := u.statusConditionSet(ramen.DRClusterValidated, metav1.ConditionFalse, reason, err.Error()); err1 != nil {
+		return err1
```

Comment:
Maybe we shouldn't lose the original error.  How about if you return something like:
`return fmt.Errorf("Original err %v. New err %v, err, err1)`